In [1]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS
import string
from heapq import nlargest



with open('spaceX_DP.txt', 'r') as file: # reading the file content
    text = file.read()

print(text[:100]) # checking the data


Cape Canaveral this is not.But here, down toward the coast, on a spit of land past the Border Patrol


In [2]:
print(string.punctuation, STOP_WORDS)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ {'’d', 'him', 'fifty', '‘s', 'could', 'became', 'front', 'take', 'after', 'who', 'full', "'ll", 'hereupon', 'third', 'nor', 'yours', 'whenever', 'just', 'i', 'almost', 'as', 'then', 'onto', 'would', 'also', 'whereupon', 'put', 'everything', 'five', 'whether', 'how', 'name', 'his', 'therein', "'m", 'being', 'may', 'next', 'does', 'each', 'somewhere', 'among', 'thus', '‘re', 'where', 'someone', 'empty', 'unless', 'me', 'has', 'well', 'except', 'since', 'until', '‘ll', 'she', "n't", 'thereupon', 're', 'up', 'what', 'mostly', 'behind', 'none', 'afterwards', 'two', 'anywhere', 'were', 'thru', 'something', 'once', 'make', 'anyhow', 'nothing', 'thereby', 'get', '‘d', 'an', 'most', 'somehow', 'another', 'alone', 'hers', 'become', 'already', 'such', 'keep', 'ca', 'own', 'whatever', 'any', 'per', 'done', 'more', 'must', 'those', 'beside', "'ve", 'off', 'via', 'is', 'often', 'along', 'around', 'why', 'every', 'here', 'others', '’ve', 'n’t', 'our', 'therefore', 'th

In [3]:
nlp = spacy.load('en_core_web_sm')  # loading the English model
doc = nlp(text)
unwanted_chars = string.punctuation + "“”‘’—"  # Adding smart quotes and em dash

tokens = []

# tokenizing the text and remove stopwords & punctuations

for w in doc:
# checking for a stopword, punctuation, and allowing only alphabetic characters
#because I have seen numbers & other special chars in between the tokens
    if not w.is_stop and w.text not in unwanted_chars and w.text.isalpha():
        tokens.append(w.text)


print(tokens[:200])

['Cape', 'Canaveral', 'coast', 'spit', 'land', 'past', 'Border', 'Patrol', 'checkpoint', 'Rio', 'Grande', 'meets', 'Gulf', 'Mexico', 'spaceship', 'assembled', 'State', 'Highway', 'dead', 'ends', 'sea', 'Towering', 'stainless', 'steel', 'shiny', 'looks', 'like', 'surreal', 'sculpture', 'amid', 'cactuses', 'yucca', 'relentless', 'South', 'Texas', 'sun', 'built', 'factory', 'open', 'roadside', 'attraction', 'drawing', 'gawkers', 'area', 'remote', 'county', 'trucks', 'drinking', 'water', 'month', 'live', 'nearby', 'coming', 'Elon', 'Musk', 'latest', 'creation', 'prototype', 'called', 'Starship', 'hopes', 'day', 'carry', 'people', 'dozens', 'moon', 'Mars', 'Musk', 'presentation', 'Saturday', 'said', 'goal', 'building', 'rapidly', 'reusable', 'spacecraft', 'lead', 'fulfillment', 'ultimate', 'goal', 'creating', 'city', 'Mars', 'need', 'pull', 'improbable', 'feat', 'building', 'private', 'commercial', 'spaceport', 'local', 'elected', 'official', 'called', 'mind', 'boggling', 'juxtaposition', '

In [4]:
print(tokens[:500]) #just making sure

['Cape', 'Canaveral', 'coast', 'spit', 'land', 'past', 'Border', 'Patrol', 'checkpoint', 'Rio', 'Grande', 'meets', 'Gulf', 'Mexico', 'spaceship', 'assembled', 'State', 'Highway', 'dead', 'ends', 'sea', 'Towering', 'stainless', 'steel', 'shiny', 'looks', 'like', 'surreal', 'sculpture', 'amid', 'cactuses', 'yucca', 'relentless', 'South', 'Texas', 'sun', 'built', 'factory', 'open', 'roadside', 'attraction', 'drawing', 'gawkers', 'area', 'remote', 'county', 'trucks', 'drinking', 'water', 'month', 'live', 'nearby', 'coming', 'Elon', 'Musk', 'latest', 'creation', 'prototype', 'called', 'Starship', 'hopes', 'day', 'carry', 'people', 'dozens', 'moon', 'Mars', 'Musk', 'presentation', 'Saturday', 'said', 'goal', 'building', 'rapidly', 'reusable', 'spacecraft', 'lead', 'fulfillment', 'ultimate', 'goal', 'creating', 'city', 'Mars', 'need', 'pull', 'improbable', 'feat', 'building', 'private', 'commercial', 'spaceport', 'local', 'elected', 'official', 'called', 'mind', 'boggling', 'juxtaposition', '

In [5]:
# getting the word frequencies
word_frequencies = {}
for t in tokens:
    if t not in word_frequencies:
        word_frequencies[t] = 1
    else:
        word_frequencies[t] += 1


In [6]:
maximum_frequency = max(word_frequencies.values()) # calculating the maximum frequency

for wrd in word_frequencies:
    word_frequencies[wrd] = word_frequencies[wrd] / maximum_frequency # normalizing the word frequencies


print(word_frequencies)


{'Cape': 0.03557312252964427, 'Canaveral': 0.02766798418972332, 'coast': 0.015810276679841896, 'spit': 0.007905138339920948, 'land': 0.05928853754940711, 'past': 0.03162055335968379, 'Border': 0.07114624505928854, 'Patrol': 0.011857707509881422, 'checkpoint': 0.019762845849802372, 'Rio': 0.03557312252964427, 'Grande': 0.03557312252964427, 'meets': 0.003952569169960474, 'Gulf': 0.023715415019762844, 'Mexico': 0.02766798418972332, 'spaceship': 0.03162055335968379, 'assembled': 0.015810276679841896, 'State': 0.039525691699604744, 'Highway': 0.05533596837944664, 'dead': 0.007905138339920948, 'ends': 0.003952569169960474, 'sea': 0.043478260869565216, 'Towering': 0.003952569169960474, 'stainless': 0.023715415019762844, 'steel': 0.03557312252964427, 'shiny': 0.003952569169960474, 'looks': 0.011857707509881422, 'like': 0.1383399209486166, 'surreal': 0.007905138339920948, 'sculpture': 0.003952569169960474, 'amid': 0.007905138339920948, 'cactuses': 0.003952569169960474, 'yucca': 0.00790513833992

In [7]:
# Tokenizing the sentences
tkn_sentences = [l.text.lower() for l in doc.sents] # making it lower case for my convinence
print(tkn_sentences[:2]) #checking

['cape canaveral this is not.', 'but here, down toward the coast, on a spit of land past the border patrol checkpoint, where the rio grande meets the gulf of mexico, there is a spaceship being assembled off state highway 4 just before it dead-ends into the sea.']


In [8]:
print(tkn_sentences[2])

towering and stainless-steel shiny, it looks like a surreal sculpture amid the cactuses, yucca and relentless south texas sun.


In [9]:
# calculating the sentence scores
sentence_scores = {}
for s in tkn_sentences:
    for word in s.split():  # getting list of words in the sentence
        if word in word_frequencies:  # checking for the word in the frequency dictionary & updating the frequency
            if s not in sentence_scores:
                sentence_scores[s] = word_frequencies[word]
            else:
                sentence_scores[s] += word_frequencies[word]




In [10]:
print(sentence_scores)

{'but here, down toward the coast, on a spit of land past the border patrol checkpoint, where the rio grande meets the gulf of mexico, there is a spaceship being assembled off state highway 4 just before it dead-ends into the sea.': 0.26877470355731226, 'towering and stainless-steel shiny, it looks like a surreal sculpture amid the cactuses, yucca and relentless south texas sun.': 0.20158102766798422, 'and, because it’s being built not in a factory but out in the open, it’s become a roadside attraction, drawing gawkers to an area so remote that the county trucks in drinking water once a month to the few who live nearby.': 0.466403162055336, 'coming to see elon musk’s latest creation, a prototype called starship that he hopes will one day carry people by the dozens to the moon and mars.': 0.44664031620553357, 'musk, in a presentation here saturday, said his goal of building a “rapidly reusable spacecraft” here would lead to the fulfillment of his ultimate goal of creating “a city on mar

In [11]:
# summarizing the sentences using nlargest module
summarized_sentences = nlargest(5, sentence_scores, key=sentence_scores.get)

In [12]:
# back to strings using join
summary = ' '.join([str(snt) for snt in summarized_sentences])

In [13]:
# printing the summary
print("Summary:", summary)
print(summary)

Summary: however, the regulatory agency that is was supposed to be guiding this environmental impact process, the faa, allowed spacex.”five years after the faa issued its final environmental impact statement, the agency issued a 23-page “written re-evaluation of the 2014 final environmental impact statement for the spacex texas launch site,” on may 21, 2019, that acknowledged spacex had switched from the falcon program to the starship project that included a new “experimental test program.” tyma said the couple attended an faa public hearing in may 2013 and said spacex officials told them there would be no more than 12 launches of the falcon 9 and falcon heavy per year, and only then would highway 4 close on the way to their home and boca chica beach. in its july 17 response, the faa said spacex has proposed operation of the starship/super heavy launch vehicle at boca chica, but to do so the company “must obtain an experimental permit or launch license from the faa office of commercial

In [14]:
# calculating lengths
summary_length = len(summary.split())
original_length = len(text.split())

print("Summary length:", summary_length)
print("Original text length:",original_length)


Summary length: 373
Original text length: 20204


In [15]:
!pip install pytextrank

In [16]:
import pytextrank
tr = nlp.add_pipe("textrank")  # creating the pipeline

/usr/local/lib/python3.10/dist-packages


In [17]:
doc = nlp(text)
tx_rnk_sentences = [sent.text for sent in doc._.textrank.summary(limit_sentences=5)]

print("summary in 5 sentences:")
for sns in tx_rnk_sentences:
    print(sns)

summary in 5 sentences:
But first, he’ll need to pull off another improbable feat, building a private, commercial spaceport here, in what the top local elected official called a “mind-boggling” juxtaposition: SpaceX, one of the hottest companies in the world, led by a Silicon Valley celebrity with nearly 29 million Twitter followers, building a rocket in a border town where nearly a third of the residents live below the poverty line.
Five years ago, SpaceX started building a launchpad here, hauling in dirt by the ton, that would allow the company a measure of freedom without the restraints that come with shooting rockets off from government sites, such as Cape Canaveral or Vandenberg Air Force Base in California, where several other companies operate.
A handful of residents who live next door to SpaceX’s facilities recently received letters from SpaceX, which said the company’s footprint in the area was going to be bigger and more disruptive than originally imagined.
The letter from Sp

## Exercise 2

In [18]:
from nltk.sentiment import SentimentIntensityAnalyzer
import nltk

nltk.download('vader_lexicon') # downloading VADER sentiment analyzer

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [19]:
sia = SentimentIntensityAnalyzer()  # creating a sentiment analyzer instance
sen_srs = sia.polarity_scores(text)

compound = sen_srs['compound'] # getting the sentiment scores
negative = sen_srs['neg']
neutral = sen_srs['neu']
positive = sen_srs['pos']

print(f"Compound: {compound}")
print(f"Negative: {negative}")
print(f"Neutral: {neutral}")
print(f"Positive: {positive}")





Compound: 1.0
Negative: 0.048
Neutral: 0.874
Positive: 0.078



"""
Compound: 1.0
Negative: 0.048
Neutral: 0.874
Positive: 0.078
"""
since the combined sentiment scores of Positive & Negative is less than 0.5, the text can be considered Objective.

